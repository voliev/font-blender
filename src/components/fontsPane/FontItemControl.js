import React, { Component } from "react";
import classNames from "classnames";
import AddCircleRounded from "@material-ui/icons/AddCircleRounded";
import SyncRounded from "@material-ui/icons/SyncRounded";
import CheckCircleRounded from "@material-ui/icons/CheckCircleRounded";
import PropTypes from "prop-types";
import "./styles/FontItemControl.css";

class FontItemControl extends Component {
  getFontStateFromProps() {
    const { active, requested, selected, fetching } = this.props;
    const stateOptions = {
      selected,
      active: active && !selected,
      fetch: !active && requested && !fetching
    };

    for (const state in stateOptions) {
      if (stateOptions[state]) {
        return state;
      }
    }
  }

  getControlIconFromFontState(fontState) {
    const icons = {
      fetch: <SyncRounded fontSize="default" />,
      active: <AddCircleRounded fontSize="default" />,
      selected: <CheckCircleRounded fontSize="default" />
    };

    return icons[fontState];
  }

  getTitleAttributeFromFontState(fontState) {
    const titleAttributes = {
      fetch: "Font has not been loaded. Try again",
      active: "Add to Selected",
      selected: "Deselect font"
    };

    const title = titleAttributes[fontState] || "Fetching...";

    return title;
  }

  render() {
    const {
      family,
      requested,
      prefetched,
      fetched,
      fetching,
      selected,
      fontSelectionDeselectionHandler,
      requestFont
    } = this.props;

    const componentClasses = classNames({
      FontItemControl: true,
      fetching
    });
    const btnClasses = classNames({
      btn: true,
      "btn-round-small": true,
      "add-circle": !selected && (prefetched || fetched),
      checked: selected && !fetching,
      sync: requested && !prefetched && !fetched
    });

    const fontState = this.getFontStateFromProps();

    return (
      <div className={componentClasses}>
        <button
          className={btnClasses}
          onClick={prefetched ? fontSelectionDeselectionHandler : requestFont}
          data-family={family}
          value={family}
          title={this.getTitleAttributeFromFontState(fontState)}
        >
          {this.getControlIconFromFontState(fontState)}
        </button>
      </div>
    );
  }
}

FontItemControl.propTypes = {
  family: PropTypes.string.isRequired,
  active: PropTypes.bool.isRequired,
  requested: PropTypes.bool.isRequired,
  prefetched: PropTypes.bool.isRequired,
  fetched: PropTypes.bool.isRequired,
  fetching: PropTypes.bool.isRequired,
  selected: PropTypes.bool.isRequired,
  fontSelectionDeselectionHandler: PropTypes.func.isRequired,
  requestFont: PropTypes.func.isRequired
};

export default FontItemControl;
