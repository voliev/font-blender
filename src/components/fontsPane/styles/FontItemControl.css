.FontItemControl {
  align-self: stretch;
  grid-area: controls;
  justify-self: end;
}

.FontItemControl.fetching {
  color: transparent;
  text-shadow: var(--blur-text);
}

.FontItemControl .btn {
  color: var(--light-grey);
  margin: 0;
  transition: color 0.1s ease-out, transform 0.1s ease-out, box-shadow 0.1s;
}

.FontItemControl .btn:hover {
  transform: scale(1.2)
}

.FontItemControl .btn.add-circle:hover {
  color: var(--green);
}

.FontItemControl .btn:focus {
  outline: none;
}

.FontItemControl .btn.add-circle,
.FontItemControl .btn.sync {
  color: var(--grey);
}

.FontItemControl .btn.checked {
  color: var(--green);
}

.FontItemControl:not(.fetching) .btn.sync {
  color: var(--orange);
}

.FontItemControl:not(.fetching) .btn.sync:hover {
  transform: scale(1.2);
  transform-origin: .75rem .75rem;
}

.FontItemControl:not(.fetching) .btn.sync:active {
  transform: rotate(-.25turn);
  transform-origin: .75rem .75rem;
}
