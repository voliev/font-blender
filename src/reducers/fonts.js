import { combineReducers } from "redux";
import {
  FETCH_FONTS_DATA,
  FETCH_FONTS_DATA_FAILURE,
  FETCH_FONTS_DATA_SUCCESS,
  SET_FONTS,
  UPDATE_FONT
} from "../constants";

export const fontsDataStatus = (state = "", action) => {
  switch (action.type) {
    case FETCH_FONTS_DATA:
      return "fetching";
    case FETCH_FONTS_DATA_SUCCESS:
      return "success";
    case FETCH_FONTS_DATA_FAILURE:
      return "failure";
    default:
      return state;
  }
};

const initialState = {};

export const all = (state = initialState, action) => {
  switch (action.type) {
    case SET_FONTS:
      return action.fonts;
    case UPDATE_FONT:
      return {
        ...state,
        [action.family]: {
          ...state[action.family],
          ...action.updates
        }
      };
    default:
      return state;
  }
};

export default combineReducers({
  all,
  fontsDataStatus
});

export const getFonts = state => state.fonts.all;
export const getFont = (state, family) => state.fonts.all[family];
export const getFontsDataStatus = state => state.fonts.fontsDataStatus;
