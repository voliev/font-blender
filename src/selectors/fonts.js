import isEqual from "lodash.isequal";

export default function() {
  const memo = {
    filters: {},
    result: []
  };

  return function(fonts, filters) {
    const families = Object.keys(fonts);

    if (families.length === 0) {
      memo.filters = { ...filters };
      memo.result = [];

      return [];
    }

    if (
      (memo.result.length === 0 && families.length > 0) ||
      !isEqual(memo.filters, filters)
    ) {
      memo.filters = { ...filters };

      const { title, categories, sorting } = filters;

      const filteredFonts = families.filter(family => {
        const { category } = fonts[family];

        const titleMatch = family
          .toLowerCase()
          .includes(title.trim().toLowerCase());
        const categoryMatch = categories[category];

        return titleMatch && categoryMatch;
      });

      if (sorting.param === "alpha") {
        const sortedFonts = filteredFonts.sort((a, b) =>
          a.localeCompare(b, {}, { caseFirst: "upper" })
        );

        memo.result =
          sorting.order === "asc" ? sortedFonts : sortedFonts.reverse();
      } else {
        memo.result =
          sorting.order === "asc" ? filteredFonts : filteredFonts.reverse();
      }
    }

    return memo.result;
  };
}
