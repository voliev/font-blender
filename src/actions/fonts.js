import WebFont from "webfontloader";
import { getFonts } from "../reducers/fonts";
import {
  GOOGLE_FONTS_API_ENDPOINT,
  FETCH_FONTS_DATA,
  FETCH_FONTS_DATA_FAILURE,
  FETCH_FONTS_DATA_SUCCESS,
  SET_FONTS,
  UPDATE_FONT
} from "../constants";

export const fetchFontsData = () => ({
  type: FETCH_FONTS_DATA
});

export const fetchFontsDataSuccess = () => ({
  type: FETCH_FONTS_DATA_SUCCESS
});

export const fetchFontsDataFailure = () => ({
  type: FETCH_FONTS_DATA_FAILURE
});

export const setFonts = fonts => ({
  type: SET_FONTS,
  fonts
});

export const startFetchFontsData = () => {
  return async dispatch => {
    const apiRequestURL = `${GOOGLE_FONTS_API_ENDPOINT}?key=${
      process.env.REACT_APP_GOOGLE_FONTS_API_KEY
    }&sort=popularity`;

    dispatch(fetchFontsData());

    try {
      const res = await fetch(apiRequestURL);
      const { items } = await res.json();
      dispatch(fetchFontsDataSuccess());

      const fonts = items.reduce((result, font) => {
        result[font.family] = {
          family: font.family,
          category: font.category,
          variants: font.variants,
          subsets: font.subsets,
          fetching: false,
          active: false,
          requested: false,
          prefetched: false,
          fetched: false
        };

        return result;
      }, {});

      return dispatch(setFonts(fonts));
    } catch (e) {
      dispatch(fetchFontsDataFailure());
      // TODO: Eject and replace service worker with custom one.
      //       Implement fonts' json data caching.
      return dispatch(setFonts({}));
    }
  };
};

export const updateFont = (family, updates) => ({
  type: UPDATE_FONT,
  family,
  updates
});

export const startFetchFont = (family, charRange = undefined) => {
  return (dispatch, getState) => {
    const state = getState();
    const fonts = getFonts(state);
    const font = fonts[family];
    const fetchFull = !charRange;
    const requestString = generateRequestString(font, fetchFull);
    const options = generateRequestOptions(
      requestString,
      charRange,
      dispatch,
      state
    );

    WebFont.load(options);

    const updatedState = getState();
    const updatedFonts = getFonts(updatedState);

    // For "then" chaining in selectedFonts.js async action creator
    return Promise.resolve(updatedFonts[family]);
  };
};

export const generateRequestString = (font, fetchFull = false) => {
  const { family, subsets, variants } = font;

  if (fetchFull) {
    return `${family}:${variants.join(",")}`;
  }

  const weight = variants.includes("regular")
    ? "400"
    : parseInt(variants[0], 10).toString();
  const subset = subsets.includes("latin") ? "latin" : subsets[0];

  return `${family}:${weight}:${subset}`;
};

export const generateRequestOptions = (request, chars, dispatch) => ({
  google: {
    families: [request],
    text: chars
  },
  classes: false,
  timeout: 10000,
  fontloading: getEventListener("fontloading", dispatch, chars),
  fontactive: getEventListener("fontactive", dispatch, chars),
  fontinactive: getEventListener("fontinactive", dispatch, chars)
});

export const getEventListener = (event, dispatch, charRange) => {
  const listeners = {
    fontloading: family => {
      const updates = Object.assign(
        { fetching: true },
        charRange ? { requested: true } : {}
      );
      dispatch(updateFont(family, updates));
    },
    fontactive: family => {
      const updates = Object.assign(
        { active: true, fetching: false },
        charRange ? { prefetched: true } : { fetched: true }
      );
      dispatch(updateFont(family, updates));
    },
    fontinactive: family => {
      const updates = Object.assign(
        { fetching: false },
        charRange ? { active: false } : {}
      );
      dispatch(updateFont(family, updates));
    }
  };

  return listeners[event];
};
