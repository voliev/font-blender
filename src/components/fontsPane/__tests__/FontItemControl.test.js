import React from "react";
import { mount, shallow } from "enzyme";
import AddCircleRounded from "@material-ui/icons/AddCircleRounded";
import SyncRounded from "@material-ui/icons/SyncRounded";
import CheckCircleRounded from "@material-ui/icons/CheckCircleRounded";
import FontItemControl from "../FontItemControl";
import fonts from "../../../fixtures/fonts";

const setup = (renderFn = shallow, propsOverrides = {}) => {
  const props = {
    ...fonts["Roboto"],
    selected: false,
    fontSelectionDeselectionHandler: jest.fn(),
    requestFont: jest.fn(),
    ...propsOverrides
  };

  return {
    ...props,
    wrapper: renderFn(<FontItemControl {...props} />)
  };
};

afterEach(() => {
  jest.clearAllMocks();
});

it("should render without crashing", () => {
  const { wrapper } = setup(shallow, {
    active: true,
    requested: true,
    prefetched: true
  });

  expect(wrapper).toMatchSnapshot();
});

it("should receive corresponding props", () => {
  const {
    active,
    family,
    fetched,
    fetching,
    prefetched,
    requested,
    fontSelectionDeselectionHandler,
    requestFont,
    wrapper
  } = setup(mount);

  expect(typeof wrapper.prop("family")).toBe("string");
  expect(wrapper.prop("family")).toBe(family);

  expect(typeof wrapper.prop("active")).toBe("boolean");
  expect(wrapper.prop("active")).toBe(active);

  expect(typeof wrapper.prop("requested")).toBe("boolean");
  expect(wrapper.prop("requested")).toBe(requested);

  expect(typeof wrapper.prop("prefetched")).toBe("boolean");
  expect(wrapper.prop("prefetched")).toBe(prefetched);

  expect(typeof wrapper.prop("fetched")).toBe("boolean");
  expect(wrapper.prop("fetched")).toBe(fetched);

  expect(typeof wrapper.prop("fetching")).toBe("boolean");
  expect(wrapper.prop("fetching")).toBe(fetching);

  expect(typeof wrapper.prop("selected")).toBe("boolean");
  expect(wrapper.prop("selected")).toBeFalsy();

  expect(typeof wrapper.prop("fontSelectionDeselectionHandler")).toBe(
    "function"
  );
  expect(wrapper.prop("fontSelectionDeselectionHandler")).toBe(
    fontSelectionDeselectionHandler
  );

  expect(typeof wrapper.prop("requestFont")).toBe("function");
  expect(wrapper.prop("requestFont")).toBe(requestFont);
});

it("should call request handler on button click", () => {
  const { wrapper, requestFont } = setup(shallow, { requested: true });

  wrapper.find("button").simulate("click");

  expect(requestFont).toHaveBeenCalledTimes(1);
});

it("shoul call selection-deselection handler on button click", () => {
  const { wrapper, family, fontSelectionDeselectionHandler } = setup(shallow, {
    active: true,
    prefetched: true,
    selected: false
  });

  const event = {
    currentTarget: {
      value: family
    }
  };

  wrapper.find("button").simulate("click", event);

  expect(fontSelectionDeselectionHandler).toHaveBeenCalledTimes(1);
  expect(fontSelectionDeselectionHandler).toHaveBeenCalledWith(event);

  wrapper.setProps({ selected: true });

  fontSelectionDeselectionHandler.mockClear();

  wrapper.find("button").simulate("click", event);

  expect(fontSelectionDeselectionHandler).toHaveBeenCalledTimes(1);
  expect(fontSelectionDeselectionHandler).toHaveBeenCalledWith(event);
});

it("should set the corresponding icon", () => {
  const { wrapper } = setup(mount);

  expect(wrapper.find(AddCircleRounded).length).toBe(0);
  expect(wrapper.find(SyncRounded).length).toBe(0);
  expect(wrapper.find(CheckCircleRounded).length).toBe(0);

  wrapper.setProps({
    active: false,
    requested: true
  });

  expect(wrapper.find(AddCircleRounded).length).toBe(0);
  expect(wrapper.find(SyncRounded).length).toBe(1);
  expect(wrapper.find(CheckCircleRounded).length).toBe(0);

  wrapper.setProps({
    active: true,
    prefetched: true
  });

  expect(wrapper.find(AddCircleRounded).length).toBe(1);
  expect(wrapper.find(SyncRounded).length).toBe(0);
  expect(wrapper.find(CheckCircleRounded).length).toBe(0);

  wrapper.setProps({ selected: true });

  expect(wrapper.find(AddCircleRounded).length).toBe(0);
  expect(wrapper.find(SyncRounded).length).toBe(0);
  expect(wrapper.find(CheckCircleRounded).length).toBe(1);
});

it("should set the corresponding title attribute", () => {
  const { wrapper } = setup();

  expect(wrapper.find("button").prop("title")).toBe("Fetching...");

  wrapper.setProps({
    active: true,
    prefetched: true
  });

  expect(wrapper.find("button").prop("title")).toBe("Add to Selected");

  wrapper.setProps({ selected: true });

  expect(wrapper.find("button").prop("title")).toBe("Deselect font");

  wrapper.setProps({ selected: false });

  expect(wrapper.find("button").prop("title")).toBe("Add to Selected");
});

it("should set corresponding component classes", () => {
  const { wrapper } = setup();

  expect(wrapper.find(".FontItemControl").hasClass("fetching")).toBeFalsy();

  wrapper.setProps({ fetching: true });

  expect(wrapper.find(".FontItemControl").hasClass("fetching")).toBeTruthy();
});

it("should set corresponding button classes", () => {
  const { wrapper } = setup(shallow, { requested: true });

  expect(wrapper.find(".btn").hasClass("add-circle")).toBeFalsy();
  expect(wrapper.find(".btn").hasClass("checked")).toBeFalsy();
  expect(wrapper.find(".btn").hasClass("sync")).toBeTruthy();

  wrapper.setProps({ prefetched: true });

  expect(wrapper.find(".btn").hasClass("add-circle")).toBeTruthy();
  expect(wrapper.find(".btn").hasClass("checked")).toBeFalsy();
  expect(wrapper.find(".btn").hasClass("sync")).toBeFalsy();

  wrapper.setProps({ selected: true });

  expect(wrapper.find(".btn").hasClass("add-circle")).toBeFalsy();
  expect(wrapper.find(".btn").hasClass("checked")).toBeTruthy();
  expect(wrapper.find(".btn").hasClass("sync")).toBeFalsy();

  wrapper.setProps({ selected: false });

  expect(wrapper.find(".btn").hasClass("add-circle")).toBeTruthy();
  expect(wrapper.find(".btn").hasClass("checked")).toBeFalsy();
  expect(wrapper.find(".btn").hasClass("sync")).toBeFalsy();
});

it("should set corresponding button click handler", () => {
  const { fontSelectionDeselectionHandler, requestFont, wrapper } = setup(
    shallow
  );

  expect(wrapper.find(".btn").prop("onClick")).toBe(requestFont);

  wrapper.setProps({ prefetched: true });

  expect(wrapper.find(".btn").prop("onClick")).toBe(
    fontSelectionDeselectionHandler
  );
});
