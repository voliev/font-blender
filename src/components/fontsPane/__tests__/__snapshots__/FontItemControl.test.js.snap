// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render without crashing 1`] = `
<div
  className="FontItemControl"
>
  <button
    className="btn btn-round-small add-circle"
    data-family="Roboto"
    onClick={[MockFunction]}
    title="Add to Selected"
    value="Roboto"
  >
    <AddCircleRoundedIcon
      fontSize="default"
    />
  </button>
</div>
`;
